### PR TITLE
Add comments to lib/Types.cpp

### PR DESCRIPTION
**Pull request by dneto0**

### CODE REVIEW

File: lib/Types.cpp [hunk 1]
```diff
@@ -214,42 +217,70 @@ Type *ExtractSmallerStructField(const DataLayout &DL, Type *Ty) {
...
+//   level of nesting "matches" the other, then return the aggregate or vector.
+//   They both have the same machine byte address, but we prefer the larger
+//   type.  Here two types "match" if they're the same, or if they are
+//   scalar numeric types (integer or float0 with the same bit size.
+// - If one type is a struct and is larger than the other type, but its
+//   smallest member is smaller than the other type, then return
+//   that member.
+// - Otherwise, if one type is smaller than the other, return the smaller one.
+// - When the sizes are the same, prefer the packed structure, if it exists.
 Type *SmallerTypeNotAliasing(const DataLayout &DL, Type *TyA, Type *TyB) {
```
Comment (dneto0) [inline lib/Types.cpp:233]: This method is possibly  misnamed.  Perhaps it should be PreferredTypeNotAliasing or something like that.

It's not always the smaller type.

File: lib/Types.cpp [hunk 2]
```diff
@@ -214,42 +217,70 @@ Type *ExtractSmallerStructField(const DataLayout &DL, Type *Ty) {
...
 }
 
+// Returns a preferred type chosen from two types.
+// In particular:
+// - If either is nullptr, return the other.
+// - When one is an aggregate or vector whose first nested member at some
+//   level of nesting "matches" the other, then return the aggregate or vector.
+//   They both have the same machine byte address, but we prefer the larger
+//   type.  Here two types "match" if they're the same, or if they are
+//   scalar numeric types (integer or float0 with the same bit size.
```
Comment (rjodinchr) [inline lib/Types.cpp:227]:
```suggestion
//   scalar numeric types (integer or float with the same bit size.
```

**Review by rjodinchr — APPROVED** [review]

File: lib/Types.cpp (hunk 2 above)
Comment (alan-baker) [reply lib/Types.cpp:227]: I think that was meant to be a closing parenthesis.

```suggestion
//   scalar numeric types (integer or float) with the same bit size.
```

File: lib/Types.cpp [hunk 3]
```diff
@@ -214,42 +217,70 @@ Type *ExtractSmallerStructField(const DataLayout &DL, Type *Ty) {
...
+      return TyA;
+    }
+    if (BitcastUtils::FindAliasingContainedType(TyB, TyA, Steps, PerfectMatch,
+                                                DL, true)) {
+      // Prefer the composite type.
+      return TyB;
+    }
   }
 
+  // Handle more cases where TyA !=TyB
```
Comment (alan-baker) [inline lib/Types.cpp:257]:
```suggestion
  // Handle more cases where TyA != TyB
```

File: lib/Types.cpp [hunk 4]
```diff
@@ -269,6 +300,8 @@ Type *SmallerTypeNotAliasing(const DataLayout &DL, Type *TyA, Type *TyB) {
   return TyA;
 }
 
+// Returns the type inferred by inspecting the users of |v|.
```
Comment (alan-baker) [inline lib/Types.cpp:303]:
```suggestion
// Returns the type of |v| inferred by inspecting its users.
```

**Review by alan-baker — APPROVED** [review]
Nits

File: lib/Types.cpp [hunk 5]
```diff
@@ -214,42 +217,70 @@ Type *ExtractSmallerStructField(const DataLayout &DL, Type *Ty) {
...
 }
 
+// Returns a preferred type chosen from two types.
+// In particular:
+// - If either is nullptr, return the other.
+// - When one is an aggregate or vector whose first nested member at some
+//   level of nesting "matches" the other, then return the aggregate or vector.
+//   They both have the same machine byte address, but we prefer the larger
+//   type.  Here two types "match" if they're the same, or if they are
+//   scalar numeric types (integer or float with the same bit size.
```
Comment (alan-baker) [inline lib/Types.cpp:227]: In applying suggestions, I think the closing parenthesis was lost.

**Pull request merged**